### PR TITLE
ClientModel: Add retry and auth policies to System.ClientModel

### PR DESCRIPTION
**Pull request by annelo-msft**
This PR adds authentication and retry policies to System.ClientModel.  It adds the retry policy into the default ClientPipeline.

It includes types:
- `ClientRetryPolicy`
- `ApiKeyAuthenticationPolicy`

### CODE REVIEW

File: sdk/core/System.ClientModel/api/System.ClientModel.net6.0.cs
Comment (annelo-msft) [inline sdk/core/System.ClientModel/api/System.ClientModel.net6.0.cs:89]: TODO: is it easy to remove MessagDelay in favor of configurable properties in this constructor?

File: sdk/core/System.ClientModel/src/Pipeline/KeyCredentialAuthenticationPolicy.cs
Comment (annelo-msft) [inline sdk/core/System.ClientModel/src/Pipeline/KeyCredentialAuthenticationPolicy.cs:11]: TODO: Change name of file

File: sdk/core/System.ClientModel/src/Pipeline/KeyCredentialAuthenticationPolicy.cs
Comment (annelo-msft) [inline sdk/core/System.ClientModel/src/Pipeline/KeyCredentialAuthenticationPolicy.cs:38]: TODO: Add tests for query policy instances

File: sdk/core/System.ClientModel/src/Pipeline/KeyCredentialAuthenticationPolicy.cs
Comment (annelo-msft) [inline sdk/core/System.ClientModel/src/Pipeline/KeyCredentialAuthenticationPolicy.cs:47]: TODO: Add tests for basic and bearer policy instances

File: sdk/core/System.ClientModel/src/Pipeline/PipelineMessageDelay.cs
Comment (annelo-msft) [inline sdk/core/System.ClientModel/src/Pipeline/PipelineMessageDelay.cs:9]: TODO: evaluate how easy it is to remove this type

File: sdk/core/System.ClientModel/tests/Pipeline/ClientPipelineFunctionalTests.cs
Comment (annelo-msft) [inline sdk/core/System.ClientModel/tests/Pipeline/ClientPipelineFunctionalTests.cs:242]: TODO: Fix this test

File: sdk/core/System.ClientModel/tests/Pipeline/ClientPipelineFunctionalTests.cs
Comment (annelo-msft) [inline sdk/core/System.ClientModel/tests/Pipeline/ClientPipelineFunctionalTests.cs:351]: TODO: Fix this test

File: sdk/core/System.ClientModel/tests/Pipeline/ClientPipelineFunctionalTests.cs
Comment (annelo-msft) [inline sdk/core/System.ClientModel/tests/Pipeline/ClientPipelineFunctionalTests.cs:433]: TODO: Fix this test

File: sdk/core/System.ClientModel/api/System.ClientModel.net6.0.cs
Comment (annelo-msft) [reply sdk/core/System.ClientModel/api/System.ClientModel.net6.0.cs:89]: Investigating here: https://github.com/Azure/azure-sdk-for-net/pull/41358

Comment (annelo-msft) [reply sdk/core/System.ClientModel/api/System.ClientModel.net6.0.cs]: Addressed

File: sdk/core/System.ClientModel/README.md
Comment (annelo-msft) [inline sdk/core/System.ClientModel/README.md:41]: TODO: Update this

File: sdk/core/System.ClientModel/src/Internal/CancellationHelper.cs
Comment (annelo-msft) [inline sdk/core/System.ClientModel/src/Internal/CancellationHelper.cs:10]: Updating to more closely match the Azure.Core version of this file.

File: sdk/core/System.ClientModel/tests/Pipeline/ClientPipelineFunctionalTests.cs
Comment (annelo-msft) [inline sdk/core/System.ClientModel/tests/Pipeline/ClientPipelineFunctionalTests.cs:437]: TODO: fix this test on net462

File: sdk/core/System.ClientModel/src/Pipeline/ApiKeyAuthenticationPolicy.cs
Comment (annelo-msft) [inline sdk/core/System.ClientModel/src/Pipeline/ApiKeyAuthenticationPolicy.cs:111]: Tracking with https://github.com/Azure/azure-sdk-for-net/issues/41374

File: sdk/core/System.ClientModel/src/Pipeline/ApiKeyAuthenticationPolicy.cs
Comment (annelo-msft) [inline sdk/core/System.ClientModel/src/Pipeline/ApiKeyAuthenticationPolicy.cs:109]: It's unclear what the correct logic is here with regard to escaping the key value and efficiently implementing idempotency for retries ... I'll look into whether we have an implementation of this policy in Azure.Core today, but if we don't, we can add this in a later version per expediency.

File: sdk/core/System.ClientModel/src/Pipeline/KeyCredentialAuthenticationPolicy.cs
Comment (annelo-msft) [reply sdk/core/System.ClientModel/src/Pipeline/KeyCredentialAuthenticationPolicy.cs]: Tracking with https://github.com/Azure/azure-sdk-for-net/issues/41374

File: sdk/core/System.ClientModel/tests/Pipeline/ClientPipelineFunctionalTests.cs
Comment (annelo-msft) [reply sdk/core/System.ClientModel/tests/Pipeline/ClientPipelineFunctionalTests.cs]: Tracking here: https://github.com/Azure/azure-sdk-for-net/pull/41325#discussion_r1456628977

File: sdk/core/System.ClientModel/README.md
Comment (KrzysztofCwalina) [inline sdk/core/System.ClientModel/README.md:58]: there does not seem to be a reason to use UriBuilder here. i..e just do request.Uri = new Uri(...);

File: sdk/core/System.ClientModel/api/System.ClientModel.net6.0.cs
Comment (KrzysztofCwalina) [inline sdk/core/System.ClientModel/api/System.ClientModel.net6.0.cs:58]: I am not sure if we want to go with "queryName" as the name. Technically, the whole section in the URI is the query. By convention, not per RFC, it's typically contains several key-value pairs separated by a delimiter. And so this maybe we name this parameter "queryKeyName"?

File: sdk/core/System.ClientModel/src/Pipeline/ApiKeyAuthenticationPolicy.cs
Comment (KrzysztofCwalina) [inline sdk/core/System.ClientModel/src/Pipeline/ApiKeyAuthenticationPolicy.cs:117]: I fear this might have bugs (duplicate or override query key-value pairs). I propose we remove query support from v1.

File: sdk/core/System.ClientModel/src/Pipeline/ClientRetryPolicy.cs
Comment (KrzysztofCwalina) [inline sdk/core/System.ClientModel/src/Pipeline/ClientRetryPolicy.cs:21]: should we hide this ctor and expose shared Default policy through a static property?

File: sdk/core/System.ClientModel/src/Pipeline/ApiKeyAuthenticationPolicy.cs
Comment (christothes) [reply sdk/core/System.ClientModel/src/Pipeline/ApiKeyAuthenticationPolicy.cs:109]: Since this is a static key-based policy, I think the correct way to handle retry idempotency would be to only add this policy as a per-call rather than pre-retry policy.

File: sdk/core/System.ClientModel/src/Pipeline/ClientRetryPolicy.cs
Comment (JoshLove-msft) [inline sdk/core/System.ClientModel/src/Pipeline/ClientRetryPolicy.cs:160]: Why does this need to be suffixed with Core?

File: sdk/core/System.ClientModel/src/Pipeline/ClientRetryPolicy.cs
Comment (JoshLove-msft) [inline sdk/core/System.ClientModel/src/Pipeline/ClientRetryPolicy.cs:182]: We need to apply a jitter - https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core/src/DelayStrategy.cs#L84-L93

File: sdk/core/System.ClientModel/src/Pipeline/ClientRetryPolicy.cs
Comment (JoshLove-msft) [inline sdk/core/System.ClientModel/src/Pipeline/ClientRetryPolicy.cs:138]: nit: combine ShouldRetry and ShouldRetryAsync into something like ShouldRetrySyncOrAsync.

File: sdk/core/System.ClientModel/src/Pipeline/ClientRetryPolicy.cs
Comment (JoshLove-msft) [inline sdk/core/System.ClientModel/src/Pipeline/ClientRetryPolicy.cs:39]: Does adding `Try` to this and the other variables add more context?

File: sdk/core/System.ClientModel/src/Pipeline/ApiKeyAuthenticationPolicy.cs
Comment (JoshLove-msft) [inline sdk/core/System.ClientModel/src/Pipeline/ApiKeyAuthenticationPolicy.cs:38]: Is this comment still relevant?

File: sdk/core/System.ClientModel/src/Pipeline/ClientRetryPolicy.cs
Comment (JoshLove-msft) [inline sdk/core/System.ClientModel/src/Pipeline/ClientRetryPolicy.cs:197]: But aren't we already adding the MessageClassifier in _this_ release?

Comment (JoshLove-msft) [reply sdk/core/System.ClientModel/src/Pipeline/ClientRetryPolicy.cs:197]: Nevermind, I see that MessageClassifier doesn't yet have the Retriable API.

File: sdk/core/System.ClientModel/src/Internal/CancellationHelper.cs
Comment (JoshLove-msft) [reply sdk/core/System.ClientModel/src/Internal/CancellationHelper.cs:10]: Just for my own understanding, why aren't we using shared source for these files? I assume we are not planning to maintain both copies.

File: sdk/core/System.ClientModel/src/Pipeline/ClientRetryPolicy.cs
Comment (JoshLove-msft) [reply sdk/core/System.ClientModel/src/Pipeline/ClientRetryPolicy.cs:21]: What about applying a default value to the `maxRetries` parameter in the other constructor and just removing this?

File: sdk/core/System.ClientModel/api/System.ClientModel.net6.0.cs
Comment (JoshLove-msft) [inline sdk/core/System.ClientModel/api/System.ClientModel.net6.0.cs:90]: Are we planning to expose a public `GetNextDelay`? Otherwise, why do we need `Core` suffix?

File: sdk/core/System.ClientModel/src/Internal/CancellationHelper.cs
Comment (annelo-msft) [reply sdk/core/System.ClientModel/src/Internal/CancellationHelper.cs:10]: Yeah, it's a good question.  The current shared source files are in an Azure namespace, and our goal in System.ClientModel is to remove all Azure branding.  I've filed https://github.com/Azure/azure-sdk-for-net/issues/41035 to track removing this duplication -- once ClientModel GA's, we can either move Core and existing client to use the unbranded shared type (this may take some coordination with the generator), look at having the Azure.Core type use implementations in ClientModel, or making some of the types public in ClientModel.  We didn't want solving this issue now to add time to ClientModel release timelines.

File: sdk/core/System.ClientModel/src/Pipeline/ApiKeyAuthenticationPolicy.cs
Comment (annelo-msft) [reply sdk/core/System.ClientModel/src/Pipeline/ApiKeyAuthenticationPolicy.cs:109]: @KrzysztofCwalina had suggested a design that enforces the pipeline position for policies - I don't have an issue tracking it yet but will file one.

File: sdk/core/System.ClientModel/src/Pipeline/ClientRetryPolicy.cs
Comment (annelo-msft) [reply sdk/core/System.ClientModel/src/Pipeline/ClientRetryPolicy.cs:182]: Cool, I'll add it in.  I had wanted to simplify but if it's required and we don't expose the configuration value for now, I don't think it adds unnecessary complexity.  Thanks for flagging this!

File: sdk/core/System.ClientModel/src/Pipeline/ClientPipeline.cs
Comment (annelo-msft) [reply sdk/core/System.ClientModel/src/Pipeline/ClientPipeline.cs:93]: Good catch - I hadn't updated all of them in the feature branch and copied this in without thinking :)

File: sdk/core/System.ClientModel/src/Pipeline/ApiKeyAuthenticationPolicy.cs
Comment (annelo-msft) [reply sdk/core/System.ClientModel/src/Pipeline/ApiKeyAuthenticationPolicy.cs:38]: Not if we cut the feature :)  I'll remove it either way

File: sdk/core/System.ClientModel/src/Pipeline/ClientRetryPolicy.cs
Comment (annelo-msft) [reply sdk/core/System.ClientModel/src/Pipeline/ClientRetryPolicy.cs:197]: Yep, that's what the comment was addressing.  I'm interested in seeing if we can separate retry classifiers from error classifiers when we add this to the design, but we can address it later I think.

File: sdk/core/System.ClientModel/src/Pipeline/ClientRetryPolicy.cs
Comment (annelo-msft) [reply sdk/core/System.ClientModel/src/Pipeline/ClientRetryPolicy.cs:192]: Interesting!  I think I had just copied this in from Azure.Core, but I can look at it again to see if I moved it over incorrectly or if it's an existing issue.

File: sdk/core/System.ClientModel/api/System.ClientModel.net6.0.cs
Comment (annelo-msft) [reply sdk/core/System.ClientModel/api/System.ClientModel.net6.0.cs:90]: I think I have GetNextDelay as an internal method, and wanted to preserve the ability to add functionality that end-users couldn't override if needed by using the Template method.  I can revisit the Template method docs to see if it makes sense to expose Core if you don't expose the outer method publicly.

File: sdk/core/System.ClientModel/src/Pipeline/ApiKeyAuthenticationPolicy.cs
Comment (annelo-msft) [reply sdk/core/System.ClientModel/src/Pipeline/ApiKeyAuthenticationPolicy.cs:117]: Tracking adding it back in https://github.com/Azure/azure-sdk-for-net/issues/41391

File: sdk/core/System.ClientModel/api/System.ClientModel.net6.0.cs
Comment (annelo-msft) [reply sdk/core/System.ClientModel/api/System.ClientModel.net6.0.cs:58]: https://github.com/Azure/azure-sdk-for-net/issues/41391

File: sdk/core/System.ClientModel/src/Pipeline/ApiKeyAuthenticationPolicy.cs
Comment (annelo-msft) [reply sdk/core/System.ClientModel/src/Pipeline/ApiKeyAuthenticationPolicy.cs:109]: Tracking with https://github.com/Azure/azure-sdk-for-net/issues/41392

File: sdk/core/System.ClientModel/src/Pipeline/ClientRetryPolicy.cs
Comment (annelo-msft) [reply sdk/core/System.ClientModel/src/Pipeline/ClientRetryPolicy.cs:205]: Addressed

File: sdk/core/System.ClientModel/src/Pipeline/ClientRetryPolicy.cs
Comment (annelo-msft) [reply sdk/core/System.ClientModel/src/Pipeline/ClientRetryPolicy.cs:26]: Doing both

Comment (annelo-msft) [reply sdk/core/System.ClientModel/src/Pipeline/ClientRetryPolicy.cs]: Addressed

File: sdk/core/System.ClientModel/src/Pipeline/ClientRetryPolicy.cs
Comment (annelo-msft) [reply sdk/core/System.ClientModel/src/Pipeline/ClientRetryPolicy.cs:203]: Actually, @JoshLove-msft, it looks like adding jitter into the ClientModel retry policy has broken Azure.Core tests in the feature branch where I'm integrating Azure.Core and System.ClientModel.  So, two questions here:
1. Can you help me understand the jitter requirement and whether this is a hard requirement or a "value add" that we could postpone to a later release?
2. It may be that I can find a better design that enables adding it without regressing Azure.Core functionality -- if it's required, I'll add an issue to track adding that in so it doesn't block this PR.

Thanks!

File: sdk/core/System.ClientModel/src/Pipeline/ApiKeyAuthenticationPolicy.cs
Comment (christothes) [reply sdk/core/System.ClientModel/src/Pipeline/ApiKeyAuthenticationPolicy.cs]: That guidance makes sense, but I'd call the key rotation on retry an edge case. Most intentional key rotations are done with overlap between the validity of the old and new keys. Otherwise you have a race between rotation and key update to avoid an outage.

File: sdk/core/System.ClientModel/src/Pipeline/ClientRetryPolicy.cs
Comment (annelo-msft) [reply sdk/core/System.ClientModel/src/Pipeline/ClientRetryPolicy.cs:203]: Tracking with https://github.com/Azure/azure-sdk-for-net/issues/41393

File: sdk/core/System.ClientModel/src/Pipeline/ClientRetryPolicy.cs
Comment (annelo-msft) [reply sdk/core/System.ClientModel/src/Pipeline/ClientRetryPolicy.cs]: Addressed

File: sdk/core/System.ClientModel/src/Pipeline/ClientRetryPolicy.cs
Comment (annelo-msft) [reply sdk/core/System.ClientModel/src/Pipeline/ClientRetryPolicy.cs:37]: I think when I originally did this work, it took me a little while to get my head around the difference between "exceptions" and "lastException" given the way this method is structured and how the references to both of them are interleaved.  I found `allTryExceptions` and `thisTryException` a little more descriptive and helpful as I was working to integrate this functionality with Azure.Core -- I considered adding a comment but that wasn't helping me track the plot further down in the file.  That said, I don't feel super strongly about a single variable name, so if you care a lot about this one, I'm happy to rename it to something you like better -- feel free to suggest an alternative!

File: sdk/core/System.ClientModel/src/Pipeline/ClientRetryPolicy.cs
Comment (JoshLove-msft) [reply sdk/core/System.ClientModel/src/Pipeline/ClientRetryPolicy.cs:182]: I assume it broke the tests because the tests were assuming that there were deterministic delays. I don't think any functionality has been regressed. I think having jitter is probably a requirement given the trouble customers can get into without having jitter applied. Here is a [good article](https://aws.amazon.com/builders-library/timeouts-retries-and-backoff-with-jitter/) on the importance of jitter.

File: sdk/core/System.ClientModel/src/Pipeline/ClientRetryPolicy.cs
Comment (annelo-msft) [reply sdk/core/System.ClientModel/src/Pipeline/ClientRetryPolicy.cs:192]: It looks like this is [the same implementation we have in Azure.Core today](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core/src/Pipeline/RetryPolicy.cs#L173-L176) ... @christothes, do you have a suggestion for how we would want to modify this code?  Any risk that we would start throwing a new exception where we weren't before, or is it just that we would be correctly handling a CancellationToken by changing this, where we had a bug before?

Comment (christothes) [reply sdk/core/System.ClientModel/src/Pipeline/ClientRetryPolicy.cs:192]: Fixing this would match the behavior of the async path, I think. My suggestion would be something like:

```c#
if (cancellationToken.WaitHandle.WaitOne(time))
{
    throw new TaskCanceledException();
}
```

We must be missing test coverage on this also, so it would be nice to validate with some tests.

Comment (JoshLove-msft) [reply sdk/core/System.ClientModel/src/Pipeline/ClientRetryPolicy.cs:192]: We should use [CancellationHelper](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/System.ClientModel/src/Internal/CancellationHelper.cs) to make sure we are consistent with how these are thrown.

Comment (JoshLove-msft) [reply sdk/core/System.ClientModel/src/Pipeline/ClientRetryPolicy.cs:192]: We still need to check if the token was cancelled. WaitOne will return true even if the token wasn't signaled.

File: sdk/core/System.ClientModel/src/Pipeline/ClientRetryPolicy.cs
Comment (annelo-msft) [reply sdk/core/System.ClientModel/src/Pipeline/ClientRetryPolicy.cs:182]: Perhaps!  Per @christothes's comment in a Core team discussion (I think it was him...), would it make sense to change the tests separately from the change to the behavior?  If I were able to share a list of the tests in question, @JoshLove-msft, do you think you'd be able to make them appropriately tolerant to additional jitter?  Then I could pull those into the Azure.Core 2.0 feature branch and see if I can add the jitter in a non-breaking way.  Let me know if you're up for that.  In the meantime I have an issue tracking adding the jitter in that we can prioritize however is needed.

File: sdk/core/System.ClientModel/tests/Pipeline/ClientPipelineFunctionalTests.cs
Comment (annelo-msft) [reply sdk/core/System.ClientModel/tests/Pipeline/ClientPipelineFunctionalTests.cs]: Thank you @AlexanderSher for finding the fix to this!! 🎉

File: sdk/core/System.ClientModel/src/Pipeline/ClientRetryPolicy.cs
Comment (annelo-msft) [reply sdk/core/System.ClientModel/src/Pipeline/ClientRetryPolicy.cs:160]: Yeah, I see the problem here, and that you flagged here: https://github.com/Azure/azure-sdk-for-net/pull/41325#discussion_r1457690403

I don't know if we want to expose the ShouldRetry method publicly, but for Azure.Core back-compat, I do think we want an extensibility point, and I think we'd rather use the template method for this kind of extensibility.  Tell you what - I'll make ShouldRetry and GetNextDelay both public so that I'm correctly following the [Template Method Pattern](https://learning.oreilly.com/library/view/framework-design-guidelines/9780135896457/ch09.xhtml#sec9_9), and we can have the conversation with architects regarding whether this is too many APIs to expose on the ClientRetryPolicy type.  It'll be interesting to look at from that perspective one way or another -- these are the implications of not having a separate type to encapsulate the delay logic.

File: sdk/core/System.ClientModel/src/Pipeline/ClientRetryPolicy.cs
Comment (annelo-msft) [reply sdk/core/System.ClientModel/src/Pipeline/ClientRetryPolicy.cs]: Addressed in ClientModel and filed https://github.com/Azure/azure-sdk-for-net/issues/41400 to track the outstanding issue in Azure.Core.

**Review by KrzysztofCwalina — APPROVED** [review]

**Pull request merged**
